### DISCUSSION TITLE
IsarError: Transaction does not match Isar instance. Make sure to use transactions from the same Isar instance.

### DISCUSSION THREAD

**Discussion by ashish-bb-16** [isar/isar#1399] [Help]
I am using Isar in my flutter module. We are using flutter module inside native Android/iOS apps. We can launch multiple flutter instances in the app. For e.g. we have a home page in flutter and another page launched from home page is in flutter. When I land on home page, the Isar db works fine, but when I navigate to other page from home page, I am getting this error. I have even tried closing the Isar instance on the previous page before navigating to the new one. Both these pages are different flutter instances launched in different flutter engines. What should I do to fix this error? Please help.

**Comment by simc** [comment]
Isar does not officially support multiple different Flutter engines.

If you could provide a sample project that leads to this issue I could prioritize the implementation 🙂

**Reply by ashish-bb-16** [reply]
Thanks @simc For quick reply. Its very difficult for me to provide a sample project due to time constraints. But multiple engines can be considered as multiple isolates. SO lets say I am doing some transactions in one isolate and then opening a new isar instance in another isolate. This issue is encountered at that time. I was first going with Hive but did not try it because I did some research and found this:
```
Hi, I also started with Hive and had the same considerations. Hive does not support concurrent access from different isolates. This is why I switched to Isar. Which does support that. More broadly the problem of cross isolate concurrency includes not only sharing data but also critical sections. That is when you need to make sure that some code is not being run in parallel. This is the problem I am trying to solve right now.
```
I have already implemented the complete project with isar and it would be very difficult to switch to another db at this point of time.

Although I am  not seeing any issues with data loss. I am still getting the data from both the Isolates , but this error is getting logged and I am unable to figure what issues this might be causing. I want to make sure there are no issues encountered in production as we have a huge user base.

**Reply by simc** [reply]
Isar definitely supports multiple isolates. I've used it from 20 isolates in parallel and there are unit tests covering this.

Without a way to reproduce it is difficult to help

**Reply by ashish-bb-16** [reply]
@simc So we have a single class that opens the Isar instance. And we have created multiple DAOs which read this single instance. This issue is encountered when there are two isolates who do not share the instance of isar. They open serparate instance and try to read or write in their individual instance. Should we change the implementation?

```dart
class FlutterCache {
  static final FlutterCache _singleton = FlutterCache._internal();

  factory FlutterCache() {
    return _singleton;
  }

  FlutterCache._internal();

  static const _tag = "FlutterCache";
  static const _dbName = "CacheDB";
  Isar? _isar;

  Isar? db() => _isar;

  Future closeDb() async {
    if (_isar != null) {
      bool closed = await _isar!.close();
      Utils.printLog(runtimeType, "Exiting page, so closed database: $closed", toFirebase: false);
    }
  }

  Future open() async {
    if (_isar == null || !(_isar?.isOpen ?? false)) {
      final dir = await getApplicationDocumentsDirectory();
      _isar = await Isar.open(
          kDebugMode
              ? [
                  PageTableSchema,
                  ImpressionTableSchema,
                  APILoggingTableSchema
                ]
              : [
                  PageTableSchema,
                  ImpressionTableSchema,
                ],
          directory: dir.path,
          name: _dbName);
    }
  }

  Future<T?> runReadTxn<T>(Future<T?>? Function()? fxn) async {
    Completer<T?> completer = Completer();
    try {
      if (fxn == null) {
        return null;
      }
      await open();
      await _isar?.txn(() async {
        completer.complete(await fxn.call());
      });
    } catch (e, s) {
      completer.complete(null);
      Utils.printLog(_tag, "Error in getting Data from DB: $e \n$s");
    }
    return completer.future;
  }

  Future<T?> runWriteTxn<T>(Future<T?>? Function()? fxn) async {
    Completer<T?> completer = Completer();
    try {
      if (fxn == null) {
        return null;
      }
      await open();
      await _isar?.writeTxn(() async {
        completer.complete(await fxn.call());
      });
    } catch (e, s) {
      completer.complete(null);
      Utils.printLog(_tag, "Error in writing data to DB: $e \n$s");
    }
    return completer.future;
  }
```

This instance is called by other classes by passing the query to the readTxn and WriteTxn function.
